xarray/core/weighted.py

```diff
@@ -100,11 +103,22 @@ def __init__(self, obj, weights):
         if not isinstance(weights, DataArray):
             raise ValueError("`weights` must be a DataArray")
 
-        if weights.isnull().any():
-            raise ValueError(
-                "`weights` cannot contain missing values. "
-                "Missing values can be replaced by `weights.fillna(0)`."
+        def _weight_check(w):
+            if np.isnan(w).any():
+                raise ValueError(
+                    "`weights` cannot contain missing values. "
+                    "Missing values can be replaced by `weights.fillna(0)`."
+                )
+            return w.data
+
+        if is_duck_dask_array(weights.data):
+            import dask.array as dsa
+
+            weights.data = dsa.map_blocks(
```

Comment:
```suggestion
            weights = weights.copy(data=dsa.map_blocks(
```
so we don't modify the original object.